src/event.rs

```diff
@@ -239,94 +221,157 @@ pub enum WindowEvent<'a> {
-        modifiers: ModifiersState,
-    },
+    PointerPress(PointerId, PointerPress),
 
-    /// An mouse button press has been received.
-    MouseInput {
-        device_id: DeviceId,
-        state: ElementState,
-        button: MouseButton,
-        #[deprecated = "Deprecated in favor of WindowEvent::ModifiersChanged"]
-        modifiers: ModifiersState,
-    },
+    ScrollStarted,
+    ScrollDiscrete(Vector<i32>),
+    ScrollSmooth(Vector<f64>),
+    ScrollEnded,
 
-    /// Touchpad pressure event.
+    /// The system window theme has changed.
     ///
-    /// At the moment, only supported on Apple forcetouch-capable macbooks.
-    /// The parameters are: pressure level (value between 0 and 1 representing how hard the touchpad
-    /// is being pressed) and stage (integer representing the click level).
-    TouchpadPressure {
-        device_id: DeviceId,
-        pressure: f32,
-        stage: i64,
-    },
+    /// Applications might wish to react to this to change the theme of the content of the window
+    /// when the system changes the window theme.
+    ///
+    /// At the moment this is only supported on Windows.
+    ThemeChanged(Theme),
+}
 
-    /// Motion on some analog axis. May report data redundant to other, more specific events.
-    AxisMotion {
-        device_id: DeviceId,
-        axis: AxisId,
-        value: f64,
-    },
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Vector<T> {
+    pub x: T,
+    pub y: T,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct KeyPress {
+    pub(crate) logical_key: Option<LogicalKey>,
+    pub(crate) scan_code: u32,
+    pub(crate) is_down: bool,
+    pub(crate) repeat_count: u32,
+    pub(crate) is_synthetic: bool,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RawKeyPress {
+    pub(crate) logical_key: Option<LogicalKey>,
+    pub(crate) scan_code: u32,
+    pub(crate) is_down: bool,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct PointerPress {
+    pub(crate) button: u8,
+    pub(crate) is_down: bool,
+    pub(crate) click_count: u32,
+}
 
-    /// Touch event has been received
-    Touch(Touch),
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RawPointerPress {
+    pub(crate) button: u8,
+    pub(crate) is_down: bool,
+}
 
+impl KeyPress {
+    pub fn logical_key(&self) -> Option<LogicalKey> {
+        self.logical_key
+    }
+    pub fn scan_code(&self) -> u32 {
+        self.scan_code
+    }
+    pub fn is_down(&self) -> bool {
+        self.is_down
+    }
+    /// The number of repeat events that have been generated from the user holding the key down. Is
+    /// `0` for the first event, and increments by `1` for every further event.
+    ///
+    /// TODO PHRASE BETTER
+    pub fn repeat_count(&self) -> u32 {
+        self.repeat_count
```

Comment:
What's the motivation for tracking this?

Reply:
The intent was to generalize keyboard repeat tracking in the same way that `click_count` generalizes double-click events, although now that I think about that more I can't think of a place where that'd be a useful generalization to make. Going with `is_repeat` for keyboard events should be fine.